port/osx/omrintrospect.c

```diff
@@ -300,9 +321,75 @@ getThreadContext(J9ThreadWalkState *state)
 
 	if (0 == ret) {
 		/* Wait for the signal handler to complete. */
-		if (0 == read(pipeFileDescriptor[0], buffer, 1)) {
+		ret = timedWait(timeout(state->deadline1));
+		if (TIMEOUT == ret) {
+			RECORD_ERROR(state, TIMEOUT, 0);
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * Utility function to calculate remaining time before the deadline elapses.
+ *
+ * @param deadline system time by which processing should have completed
+ *
+ * @return number of seconds before timeout.
+ */
+static int32_t
+timeout(int64_t deadline)
+{
+	int32_t secs = 0;
+	struct timespec spec;
+	if (-1 == clock_gettime(CLOCK_REALTIME, &spec)) {
+		return 0;
+	}
+
+	secs = (int32_t)(deadline - spec.tv_sec);
+	if (secs > 0) {
+		return secs;
+	}
+
+	return 0;
```

Comment:
Suggest the following (stricter test of `clock_gettime()`, single return, safer test that `deadline` is in the future):
```
	int32_t secs = 0;
	struct timespec spec;
	if ((0 == clock_gettime(CLOCK_REALTIME, &spec)) && (deadline > spec.tv_sec)) {
		secs = (int32_t)(deadline - spec.tv_sec);
	}

	return secs;
```